src/message_dao.py

```diff
@@ -12,27 +12,137 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Dict, Optional
+import json
+import os
+import subprocess
+import zlib
+
+from functools import lru_cache
+from typing import Dict, List, Optional
+
+from dotenv import load_dotenv
+from google.cloud.sql.connector import Connector
 from message import Message
 
+load_dotenv()
 EPOCH_HASH = 'ae9e7be4a03765456fe38287533e6446e8bbc93c'
 
 class MessageDao(object):
+    def __init__(self) -> None:
+        self._initialize_connection()
+        self._initialize_tables()
+
+    def _initialize_connection(self) -> None:
+        connector = Connector()
+        self.connection = connector.connect(
+            os.environ.get("HOST"),
+            "pymysql",
+            user = os.environ.get("USER"),
+            password = os.environ.get("PASSWORD")
+        )
+    
+    def _initialize_tables(self) -> None:
+        db_name = os.environ.get("DB")
+        if not db_name:
+            raise Exception("Missing environment variable for name of database.")
+        with self.connection.cursor() as cursor:
+            cursor.execute("CREATE DATABASE IF NOT EXISTS " + db_name)
+            self.connection.select_db(db_name)
+            # Mapping from message id to message
+            cursor.execute(
+                "CREATE TABLE IF NOT EXISTS Messages"
+                "(message_id VARCHAR(255) NOT NULL,"
+                "subject VARCHAR(255) NOT NULL,"
+                "from_ VARCHAR(255) NOT NULL,"
+                "in_reply_to VARCHAR(255),"
+                "content BLOB NOT NULL,"
+                "archive_hash VARCHAR(255),"
```

Comment:
Sorry, I should have brought this up a lot earlier, but I forgot.

more meta comment, the main thing we really need is a 
  `message_id` => `archive_hash`
mapping, and some indication of what message_id's we've uploaded.

With the `archive_hash` in hand, we can fetch the raw email from the archive.
From there we can get the subject, from, in_reply_to_, and content.
So we don't really need them to be in the DB necessarily.

The one other field we don't know from the `change_id`.
But we currently have this invariant: `change_id == sha1sum(msg_id)`.
For proper version tracking, we'll need to have this be `change_id == sha1sum(first_version_msg_id)`.

`first_version_msg_id` will be `== msg_id` for the first version of a patch (i.e. for v1's or for new patches introduced in a v2+ patch series)

So theoretically, all need to store is
```
  message_id,
  archive_hash,
  first_version_message_id,  # this might be the same as message_id for v1s, etc.
``` 

There's some debuggability benefits to storing everything in one place, but something to consider.
It also has the drawback of you needing to do more to re-populate the message fields.

For concreteness, here's how we could fetch the message contents:
```
raw_email = subprocess.check_output(['git', '-C', archive_path, 'show', f'{hash}:m'])
msg = message.parse_message_from_str(raw_email, archive_hash=hash)
```

Reply:
Changed the code to acquire messages using archive hash. Most columns ended up staying as they are needed for something other than just storing the message. However, by doing this we no longer have to store 'content', which is easily the largest column. For ease of access/replacement of column, I added a `lore_link` column that allows one to easily obtain the content of a particular row if needed.

Reply:
Keeping the other fields SGTM.

Note: the lore link is just https://lore.kernel.org/linux-kselftest/<msg-id>
So we should probably drop it, tbh.

Even when we start archiving more, we can just use https://lore.kernel.org/lkml/<msg-id> to be agnostic of which sub-list it came from (and https://lore.kernel.org/all/ is an option as well)

Reply:
I actually suggested using the lore link. I like the idea because it makes it really easy to look the content of the patch up if we are accessing the database manually for research purposes.